beluga_amcl/src/ndt_amcl_node_3d.cpp

```diff
@@ -0,0 +1,447 @@
+  covariance.coeffRef(5, 5) = get_parameter("initial_pose.covariance_yaw").as_double();
+
+  return std::make_pair(pose, covariance);
+}
+
+auto NdtAmclNode3D::get_motion_model() const -> MotionModelVariant {
+  const auto name = get_parameter("robot_model_type").as_string();
+  if (name == kDifferentialModelName || name == kDifferentialModelName) {
+    auto params = beluga::DifferentialDriveModelParam{};
+    params.rotation_noise_from_rotation = get_parameter("alpha1").as_double();
+    params.rotation_noise_from_translation = get_parameter("alpha2").as_double();
+    params.translation_noise_from_translation = get_parameter("alpha3").as_double();
+    params.translation_noise_from_rotation = get_parameter("alpha4").as_double();
+    return beluga::DifferentialDriveModel<Sophus::SE3d>{params};
+  }
+  throw std::invalid_argument(std::string("Invalid motion model: ") + name);
+}
+
+beluga::NDTSensorModel<NDTMapRepresentation> NdtAmclNode3D::get_sensor_model() const {
+  auto params = beluga::NDTModelParam3d{};
+  params.minimum_likelihood = get_parameter("minimum_likelihood").as_double();
+  params.d1 = get_parameter("d1").as_double();
+  params.d2 = get_parameter("d2").as_double();
+  const auto map_path = get_parameter("map_path").as_string();
+  RCLCPP_INFO(get_logger(), "Loading map from %s.", map_path.c_str());
+
+  return beluga::NDTSensorModel<NDTMapRepresentation>{
+      params, beluga::io::load_from_hdf5<NDTMapRepresentation>(get_parameter("map_path").as_string())};
+}
+auto NdtAmclNode3D::make_particle_filter() const -> std::unique_ptr<NdtAmclVariant> {
+  auto amcl = std::visit(
+      [this](auto motion_model, auto execution_policy) -> NdtAmclVariant {
+        auto params = beluga::AmclParams{};
+        params.update_min_d = get_parameter("update_min_d").as_double();
+        params.update_min_a = get_parameter("update_min_a").as_double();
+        params.resample_interval = static_cast<std::size_t>(get_parameter("resample_interval").as_int());
+        params.selective_resampling = get_parameter("selective_resampling").as_bool();
+        params.min_particles = static_cast<std::size_t>(get_parameter("min_particles").as_int());
+        params.max_particles = static_cast<std::size_t>(get_parameter("max_particles").as_int());
+        params.alpha_slow = get_parameter("recovery_alpha_slow").as_double();
+        params.alpha_fast = get_parameter("recovery_alpha_fast").as_double();
+        params.kld_epsilon = get_parameter("pf_err").as_double();
+        params.kld_z = get_parameter("pf_z").as_double();
+
+        auto hasher = beluga::spatial_hash<Sophus::SE3d>(
+            get_parameter("spatial_resolution_x").as_double(), get_parameter("spatial_resolution_theta").as_double());
+
+        // We don't support randomly sampling from NDT maps, so we enforce a non-adaptive filter.
+        assert(params.min_particles == params.max_particles);
+        RandomStateGenerator random_state_maker = []() { return Sophus::SE3d{}; };
```

Comment:
One thought about random sampling these maps. If sparse, sampling from a uniform over their convex hull (or some other user-defined boundary) or a few sigmas away from any given normally distributed "voxel" will often (if not almost always) yield free space. We just need an efficient way of checking that it is indeed free.

Reply:
> @serraramiro1 meta: should we have a beluga::Mcl class for this instead of hacking beluga::Amcl? Also, why NdtAmcl if it's not really adaptive?


Probably, or at least an option to disable the adaptiveness.

There's a ticket for this:
https://github.com/Ekumen-OS/beluga/issues/424

> One thought about random sampling these maps. If sparse, sampling from a uniform over their convex hull (or some other user-defined boundary) or a few sigmas away from any given normally distributed "voxel" will often (if not almost always) yield free space. We just need an efficient way of checking that it is indeed free.

Randomly sampling 3D space seems like a hail mary, to me, how likely is it that we sample a decent state from a 3D map? It get's way more unlikely. Also, rotations come into play aswell. I personally don't expect that uniformly sampling SO3, + picking free 3D translations, regardless of the sampling approach will yield good returns for realistic use cases.

Reply:
Define decent state. Likelihood function variance plays a big role here. I wouldn't outright disregard the approach, at least not without some experimentation.

FWIW it's not that you don't have a point. You do. We know this will eventually break down -- not just random sampling, but the bootstrap particle filter itself. But it'll be easier to justify an optimal transport particle filter when we get there.

Reply:
Fair.
Anyways, this won't happen as part of this PR. How should we move forward with this?